### ISSUE TITLE
OPENSSL_atexit is not thread safe

### ISSUE THREAD

**Issue by bernd-edlinger** [openssl/openssl#17490]
Hi, I noticed that OPENSSL_atexit is not thread safe, see:
https://github.com/openssl/openssl/blob/a4e01187d3648d9ce99507097400902cf21f9b55/crypto/init.c#L708-L710
here `stop_handlers` is a global variable.
that would be okay if there was only one caller, but there are three of them:

```
master:
apps/lib/vms_decc_argv.c:                                OPENSSL_atexit(cleanup_argv);
providers/implementations/rands/seeding/rand_unix.c:     OPENSSL_atexit(&cleanup_shm);
ssl/ssl_init.c:                                          OPENSSL_atexit(ssl_library_stop);
```

```
1.1.1:
crypto/rand/rand_unix.c:                OPENSSL_atexit(&cleanup_shm);
crypto/store/loader_file.c:             OPENSSL_atexit(store_file_loader_deinit);
ssl/ssl_init.c:                         OPENSSL_atexit(ssl_library_stop);
```

they don't seem to be synchronized against each other.

**Comment by paulidale** [comment]
Yeah, I was aware of this.  We don't handle parallel initialisation at all well.  I've flagged this as a bug for lack of anything better to call it.  I'm not sure it is a bug, perhaps more a limitation of the library.

**Comment by nhorman** [comment]
Given that we have provided a no-atexit as a configuration option, and are planning on removing atexit usage in 4.0 (IIRC), is there still anything to be done here?

**Comment by nhorman** [comment]
no response, moving to inactive, to be closed at the end of 3.4 dev barring further input
